src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsauditapi/services/AuditService.kt

```diff
@@ -5,25 +5,31 @@ import org.slf4j.Logger
 import org.slf4j.LoggerFactory
 import org.springframework.stereotype.Service
 import uk.gov.justice.digital.hmpps.hmppsauditapi.config.trackEvent
+import uk.gov.justice.digital.hmpps.hmppsauditapi.jpa.AuditRepository
 import uk.gov.justice.digital.hmpps.hmppsauditapi.listeners.HMPPSAuditListener.AuditEvent
 
 @Service
-class AuditService(private val telemetryClient: TelemetryClient) {
+class AuditService(private val telemetryClient: TelemetryClient, private val auditRepository: AuditRepository) {
   companion object {
     val log: Logger = LoggerFactory.getLogger(this::class.java)
   }
 
   fun audit(auditEvent: AuditEvent) {
     log.info("About to audit $auditEvent")
     telemetryClient.trackEvent("hmpps-audit", auditEvent.asMap())
+    auditRepository.save(auditEvent)
   }
 }
 
 private fun AuditEvent.asMap(): Map<String, String> {
-  var items = mapOf("what" to this.what, "when" to this.`when`)
-  this.who?.let { items = items + ("who" to it) }
-  this.service?.let { items = items + ("service" to it) }
-  this.details?.let { items = items + ("details" to it) }
-
+  var items = mutableMapOf("what" to this.what, "when" to this.`when`.toString())
+  items.addIfNotNull("operationId", this.operationId)
+  items.addIfNotNull("who", this.who)
+  items.addIfNotNull("service", this.service)
+  items.addIfNotNull("details", this.details)
   return items
```

Comment:
I think you need a `items.toMap()` here because otherwise the map is still technically mutable (i.e. you could cast it to a MutableMap and start changing it).

Reply:
yes, you're right.  I missed the comment before I merged back into main...will update